### PR TITLE
Use enum class for TLVElementType and TLVTagControl.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
This makes it easy to give them as specific size spec, which will
prevent issues with people adding out-of-range values and simplify
-Wconversion work.

 #### Problem
TLVElementType and TLVTagControl are raw enums, not enum classes, so proving that casts to 8-bit ints are safe is not trivial.

 #### Summary of Changes
Make them 8-bit-backed enum classes, so the compiler will enforce things for us.

### CODE REVIEW

File: src/lib/core/CHIPTLVData.hpp [hunk 1]
```diff
@@ -105,41 +105,41 @@
 /*
  *  @brief Specifies an anonymous TLV element, which doesn't have any tag
  */
-#define CHIP_TLV_TAG_ANONYMOUS chip::TLV::kTLVTagControl_Anonymous
+#define CHIP_TLV_TAG_ANONYMOUS chip::TLV::TLVTagControl::Anonymous
```
Comment (rwalker-apple) [inline src/lib/core/CHIPTLVData.hpp:108]: this departs from C++ towards Haskell :-)

**Review by rwalker-apple — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

**Review by mspang — APPROVED** [review]

**Review by BroderickCarlin — APPROVED** [review]

File: src/lib/core/CHIPTLV.h [hunk 2]
```diff
@@ -58,6 +58,16 @@ class PacketBuffer;
 namespace chip {
 namespace TLV {
 
+inline uint8_t operator|(TLVElementType lhs, TLVTagControl rhs)
```
Comment (gerickson) [inline src/lib/core/CHIPTLV.h:61]: Style nit: but strongly prefer white space around `operator` the operation `|` to improve readability.

**Review by gerickson — APPROVED** [review]

File: src/lib/core/CHIPTLV.h (hunk 2 above)
Comment (rwalker-apple) [reply src/lib/core/CHIPTLV.h:61]: can we make .clang-format do that?  if not, we're gonna have trouble enforcing

Comment (gerickson) [reply src/lib/core/CHIPTLV.h:60]: Agreed. I can't seem to find anything in the clang-format documentation that indicates control over this specific behavior.

Comment (bzbarsky-apple) [reply src/lib/core/CHIPTLV.h:60]: As in, whitespace between `|` and `(`?  Our clang-format configuration will go ahead and remove it if I add it in...  Maybe we could change that configuration?

**Pull request merged**